exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -138,7 +166,103 @@ public static TestKit forService(Class<? extends ServiceModule> serviceModule) {
+  /**
+   * Creates a block with the given transaction. In-pool transactions will be ignored.
+   *
+   * @return created block
+   * @see <a href="https://exonum.com/doc/version/0.10/advanced/consensus/specification/#pool-of-unconfirmed-transactions">Pool of Unconfirmed Transactions</a>
+   */
+  public Block createBlockWithTransaction(TransactionMessage transaction) {
+    return createBlockWithTransactions(transaction);
+  }
+
+  /**
+   * Creates a block with the given transaction(s). In-pool transactions will be ignored.
+   *
+   * @return created block
+   * @see <a href="https://exonum.com/doc/version/0.10/advanced/consensus/specification/#pool-of-unconfirmed-transactions">Pool of Unconfirmed Transactions</a>
+   */
+  public Block createBlockWithTransactions(TransactionMessage transaction,
+                                           TransactionMessage... transactions) {
+    return createBlockWithTransactions(asList(transaction, transactions));
+  }
+
+  /**
+   * Creates a block with the given transactions. In-pool transactions will be ignored.
+   *
+   * @return created block
+   * @see <a href="https://exonum.com/doc/version/0.10/advanced/consensus/specification/#pool-of-unconfirmed-transactions">Pool of Unconfirmed Transactions</a>
+   */
+  public Block createBlockWithTransactions(Iterable<TransactionMessage> transactions) {
+    List<TransactionMessage> messageList = ImmutableList.copyOf(transactions);
+    byte[][] transactionMessagesArr = messageList.stream()
+        .map(TransactionMessage::toBytes)
+        .toArray(byte[][]::new);
+    byte[] block = nativeCreateBlockWithTransactions(nativeHandle.get(), transactionMessagesArr);
+    return BLOCK_SERIALIZER.fromBytes(block);
+  }
+
+  /**
+   * Creates a block with all in-pool transactions.
+   *
+   * @return created block
+   * @see <a href="https://exonum.com/doc/version/0.10/advanced/consensus/specification/#pool-of-unconfirmed-transactions">Pool of Unconfirmed Transactions</a>
+   */
+  public Block createBlock() {
+    byte[] block = nativeCreateBlock(nativeHandle.get());
+    return BLOCK_SERIALIZER.fromBytes(block);
+  }
+
+  /**
+   * Returns a list of in-pool transactions that match the given predicate.
+   *
+   * @see <a href="https://exonum.com/doc/version/0.10/advanced/consensus/specification/#pool-of-unconfirmed-transactions">Pool of Unconfirmed Transactions</a>
+   */
+  public List<TransactionMessage> findTransactionsInPool(Predicate<TransactionMessage> predicate) {
+    return withSnapshot((view) -> {
+      Blockchain blockchain = Blockchain.newInstance(view);
+      MapIndex<HashCode, TransactionMessage> txMessages = blockchain.getTxMessages();
+      // As only executed transactions are stored in TxResults, it wouldn't contain in-pool
+      // transactions
+      ProofMapIndexProxy<HashCode, TransactionResult> txResults = blockchain.getTxResults();
+      List<TransactionMessage> messages = ImmutableList.copyOf(txMessages.values());
```

Comment:
> It is somewhat weird (and inefficient) that one shall go over all transactions submitted to blockchain ever to find in-pool ones. I'd submit an issue to consider providing access to transaction_pool in Blockchain (or separately).

I looked at it and core schema only returns KeySetIndex of transaction hashes of in-pool transactions, not transactions themselves. So if we want to look up transactions using predicate, that won't work. However, we could look up transactions by hashes instead.

Reply:
Why wouldn't it work? It will have to request the messages from the same txMessages map (`hashes.stream().map(txMessages::get)`), but it will do O(num_in_pool) instead of O(num_ever_sumbitted).

Reply:
Added https://github.com/exonum/exonum-java-binding/pull/850 for that

Reply:
Why this code still iterates over all messages?